src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs

```diff
@@ -422,9 +423,9 @@ private void Write(RuntimeModel.IObjectWriter writer, Action<PackageSpec, Runtim
             writer.WriteObjectStart("restore");
 
             // Preserve default sort order
-            foreach (var restoreName in _restore)
+            foreach (var restoreName in _restore.OrderBy(r => r.Key, PathUtility.GetStringComparerBasedOnOS()))
```

Comment:
Isn't this already sorted via comparer to `_restore`'s constructor?

Reply:
No, it's not anymore. Because there is no sorted version of Concurrent Set nor Concurrent Dictionary.  So now insertion is cheaper O(1) instead of O(LogN) but pay more price here O(N LogN). Everything has 2 sides, only depending when how pay the price.